### PR TITLE
lua: add request info dynamic metadata api

### LINKED ISSUE #3618: Add Lua API for accessing dynamic metadata
*Description*:

Add Lua API for accessing dynamic metadata as discussed in: https://github.com/envoyproxy/envoy/issues/2436#issuecomment-396662370

The API could be done as follows:

```lua
function envoy_on_request(request_handle)
  request_handle:requestInfo():dynamicMetadata():set("envoy.lb", "foo", "bar")
  request_handle:requestInfo():dynamicMetadata():get("envoy.lb")["foo"]
end
```
However this is still a rough design. This needs more input.

@moss94 @sahanaha

### PR DESCRIPTION
**Pull request by dio**
lua: add request info dynamic metadata API

*Description*:
This patch allows to get and set the dynamic metadata of a request via Lua.

Setting metadata,

```lua
request_handle:requestInfo():dynamicMetadata():set("envoy.lb", "foo", "bar")
```

Getting metadata,

```lua
-- returns a table
request_handle:requestInfo():dynamicMetadata():get("envoy.lb")
request_handle:requestInfo():dynamicMetadata():get("envoy.lb")["foo"]
```

*Risk Level*: Low

*Testing*: Unit and integration tests

*Docs Changes*:
- Added request info object API
- Added dynamic metadata object API

*Release Notes*:
- Added request info dynamic metadata Lua API

Fixes #3618

### CODE REVIEW

**Comment by mattklein123** [comment]
@dio is this still WIP or does this need a reviewer?

**Comment by dio** [comment]
@mattklein123 yes it is still WIP I will add docs, support for accepting a table as param the I'll ask for a review.

Two questions, 

1. Do you think this is still considered an enhancement? Since there is an effort here: https://github.com/envoyproxy/envoy/issues/2436 for making it accessible via config.
2. While this is part of the review, what do you think about the proposed API?

Thanks!

File: source/extensions/filters/common/lua/wrappers.cc [hunk 1]
```diff
@@ -125,6 +127,85 @@ int MetadataMapWrapper::luaPairs(lua_State* state) {
...
+  if (metadata_wrapper_.get() != nullptr) {
+    metadata_wrapper_.pushStack();
+  } else {
+    metadata_wrapper_.reset(DynamicMetadataMapWrapper::create(state, request_info_), true);
+  }
+  return 1;
+}
+
+int RequestInfoWrapper::luaProtocol(lua_State* state) {
+  switch (request_info_.protocol().value()) {
```
Comment (ramaraochavali) [inline source/extensions/filters/common/lua/wrappers.cc]: I have not looked at the entire code. But one quick comment here, this whole `switch` can be replaced with this [method call](https://github.com/envoyproxy/envoy/blob/master/source/common/http/utility.cc#L328)

Comment (dio) [reply source/extensions/filters/common/lua/wrappers.cc]: Gotcha! Thanks @ramaraochavali!

**Comment by mattklein123** [comment]
> Do you think this is still considered an enhancement?

@dio what do you mean? Are you asking if I think this feature is no longer necessary? TBH I don't know I haven't been tracking all of this very closely. In general I would prefer there is only a single way to do something where reasonable, but if there are good reasons to have multiple ways am open to it. Thoughts?

**Comment by ramaraochavali** [comment]
Not sure what @dio  is referring to here. But exposing Protocol Identification attributes as described in the issue https://github.com/envoyproxy/envoy/issues/3704 is definitely required for us and I think would be a reasonable addition.

**Comment by dio** [comment]
Sorry for introducing confusion. Will breakdow this PR into:

1. RequestInfo with protocol()
2. RequestInfo with dynamicMetadata()
3. Connection with secure() (whether the current conn using SSL or not).

I’ll continue this hopefully tonight. Thanks

**Comment by ramaraochavali** [comment]
@dio sure..no problem.. Thanks for doing #1 and #3. lmk if it becomes too much work for you.

**Comment by dio** [comment]
@ramaraochavali @mattklein123 

I have put PRs related to #3704:

1. `requestInfo():protocol()` in https://github.com/envoyproxy/envoy/pull/3760
2. `connection():secure()` in https://github.com/envoyproxy/envoy/pull/3761

While this PR will be layered on top of https://github.com/envoyproxy/envoy/pull/3760 if approved and merged.

Thanks.

**Comment by dio** [comment]
Closing this for now in favor of https://github.com/envoyproxy/envoy/pull/3760 for preparing the requestInfo wrapper.

Related: https://github.com/envoyproxy/envoy/pull/3761

**Comment by dio** [comment]
Resubmitted as #3800 for adding dynamic metadata API

**Pull request closed without merging**